### PR TITLE
hip-0006: live auctions

### PR DESCRIPTION
**Pull request by pinheadmz**
Requesting HIP number and community feedback on this HIP-1/Shakedex extension for live & ascending-price trustless auctions.

### CODE REVIEW

File: HIP-XXXX.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# HIP-XXXX : Trustless Live Auctions for Handshake
+
+```
+Number:  HIP-XXXX
+Title:   Trustless Live Auctions for Handshake
+Type:    Informational
+Status:  Final
+Authors: Matthew Zipkin (@pinheadmz)
+Created: 2020-06-17
```
Comment (Falci) [inline HIP-XXXX.md:9]:
```suggestion
Created: 2021-06-17
```

**Comment by Falci** [comment]
Assign the next number: `HIP-0006`

File: HIP-0006.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+## Protocol Summary
+
+### Auction OPEN
+
+The auction technically begins when a name owner FINALIZEs a transfer of their name
+to a HIP-0001 locking script. The name owner must then publicly reveal the redeem script
+for the address that owns the name, so it can be verified. The locking script prevents
+the name owner from canceling an in-progress name ownership transfer. Since there is a
+small amount of interactivity required in the protocol, the name owner must also publish
+"where" the auction will take place. The auction "location" is just some internet endpoint
```
Comment (pbkhrv) [inline HIP-0006.md:42]: "must also publish" sounds a bit too strong, could be misunderstood to be a requirement? Why not relax it to something like "it is up to the name owner to communicate the existence of this auction and to provide means for potential buyers to submit bids using mechanisms that are outside of the scope of this proposal. For example, ...."

File: HIP-0006.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+the incoming payment to the name transfer. Note that Input 0 is not signed yet,
+but it must be the current name-owning UTXO or the linked output 0 would be invalid.
+
+Note that this transaction has no change output. Because the bidder is signing
+with single-reverse, they can't specify a change output in addition to the name
+transfer output. If they were to sign with `SIGHASH_ALL` then the next step of
+the protocol would not be possible. If they signed with `SIGHASH_SINGLE` they could
+commit to a change address, but not the name transfer.
+
+Having no change output means the INPUT coin they are spending MUST BE THE VALUE
```
Comment (pbkhrv) [inline HIP-0006.md:70]: Is this caveat important enough to be explicitly mentioned again in a separate "Tradeoffs" or "Caveats" or something like that section, along with the "Seller pays the miner fee" below?

File: HIP-0006.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+```
+
+The name owner pays themselves from the fully funded transaction, and signs input 0
+with `SIGHASH_ALL`. This transaction is a valid atomic swap and can be broadcast to the network.
+Because the locking script prevents all other name actions, the auction winner can
+`FINALIZE` this name transfer any time after the 288-block lockout without any other
+obligation.
+
+Note that since the bidder creates the INPUT and the name owner creates the OUTPUT,
+the name owner MUST also pay the miner fee for this transaction. They do so by reducing
```
Comment (pbkhrv) [inline HIP-0006.md:103]: Is the "owner pays mining fee" caveat important enough to be mentioned again in a separate section?

File: HIP-0006.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+## Extra Notes
+
+### Bids must remain valid off-chain
+
+The partially-signed bid transactions are collected by the name owner and commit
+to spending a UTXO of some value V from the blockchain. In the simplest construction,
+nothing prevents the bidder from invalidating this off-chain transaction by
+double-spending the coin they used to bid. For this reason, the name owner must
+routinely check that all submitted bids are still valid. They can try finalizing
```
Comment (pbkhrv) [inline HIP-0006.md:115]: This could also be another use case for an auction platform that might take care of the auction finality issue mentioned below. Not sure if important enough to highlight?

File: HIP-0006.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+terms-of-service. In more common cases we can assume a platform with interest in
+maintaining a good reputation would award the highest bidder with the name and
+pay the name owner from output 1 of the final transaction.
+
+### Where & When
+
+Because this auction protocol is interactive, bidders need a place to send their
+partially-signed transactions where they are accessible to the name owner to finalize.
+A name owner could therefore open an auction by adding TXT records to their zone
+announcing the endpoint to send bids, along with the redeem script to prove the name
```
Comment (pbkhrv) [inline HIP-0006.md:168]: i'm hoping we can introduce a HIP to cover this use case... working on that now

File: HIP-0006.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    OP_DROP
+    <platform's public key>
+  OP_ENDIF
+  OP_CHECKSIG
+OP_ELSE
+  OP_TYPE
+  0x0a // FINALIZE
+  OP_EQUAL
+OP_ENDIF
+```
```
Comment (pbkhrv) [inline HIP-0006.md:151]: is an output locking script like this "templated" enough that something like Bob Wallet can actually detect it and mark a name as "in auction" ? (i know, a n00b question). If so, is that interesting/important enough to highlight?

Comment (tynes) [reply HIP-0006.md:151]: It would only be able to do that if it had the preimage because only the hash of the script is on chain

File: HIP-0006.md (hunk 2 above)
Comment (pinheadmz) [reply HIP-0006.md:42]: Hm, I understand your sentiment but it is actually a requirement of this protocol.

File: HIP-0006.md (hunk 3 above)
Comment (pinheadmz) [reply HIP-0006.md:70]: I'll mention both of these tradeoffs in the abstract, thanks.

File: HIP-0006.md (hunk 4 above)
Comment (pinheadmz) [reply HIP-0006.md:103]: 👍

File: HIP-0006.md (hunk 5 above)
Comment (pinheadmz) [reply HIP-0006.md:115]: Sure, will add.

File: HIP-0006.md (hunk 6 above)
Comment (pinheadmz) [reply HIP-0006.md:168]: Great!

**Comment by pinheadmz** [comment]
@pbkhrv thanks for the review! I addressed your comments and added a new commit: 82adb27e3d5afe2e089ca6b85610e8e0aa58fc75

**Comment by kurumiimari** [comment]
LGTM, no changes suggested from my end. Thanks for putting this together! A couple of thoughts:

- I think that the semi-custodial platform is almost necessary here, since it's pretty easy for both sides to grief each other without some kind of always-on service.
- I'm curious about the lightning-network-style multisigs you mention. Could you shed some light on how those might be applied here so that I can explore more on my own?

**Comment by pinheadmz** [comment]
@kurumiimari Thanks for the review!

> * I think that the semi-custodial platform is almost necessary here, since it's pretty easy for both sides to grief each other without some kind of always-on service.

True, but technically a platform isn't required for the protocol to work. I like the potential for personal, self-hosted and PRIVATE auctions. For example the owner of a super high value name like `.sex` might only invite a short list of interested parties to bid, and those bids can be submitted by encrypted email without other bidders knowing the values.

> * I'm curious about the lightning-network-style multisigs you mention. Could you shed some light on how those might be applied here so that I can explore more on my own?


I'm super glad you called me out on this ;-) After fleshing it out, I just decided to add an entire new section the HIP in this commit: https://github.com/handshake-org/HIPs/pull/23/commits/2dcfb55e9d17f319a7a6f8d2d3162118de79140e Really interested in your thoughts, and review of the outputs scripts.

File: HIP-0006.md [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+    1. `bid_lock_script` with all data inserted such that the name owner can verify it
+    matches the address in the output of `bid_lock_tx`.
+
+    2. The double-partially-signed `bid_submit_tx` (which will explicitly indicate the outpoint of the UTXO created by `bid_lock_tx`)
+
+**FINALIZE:** The name owner must finalize the atomic swap BEFORE `expiration_height`:
+
+1. Name owner chooses the highest bid and verifies that its `bid_lock_tx` is confirmed on chain,
+and that the output address is spendable by `bid_lock_script`. They construct a final TX as
+defined in [Auction FINALIZE](#auction-finalize):
```
Comment (pinheadmz) [inline HIP-0006.md:168]: The name owner should also check the TX's locktime -- although @rithvikvibhu has got me thinking that more interesting constructions like "make a limited time offer" are possible with *shorter* locktimes.
